internal/routes.go

```diff
@@ -0,0 +1,68 @@
+package internal
+
+import (
+	"encoding/binary"
+	"io"
+	"io/ioutil"
+	"time"
+)
+
+// Route is the value part of a shortcut.
+type Route struct {
+	URL  string    `json:"url"`
+	Time time.Time `json:"time"`
+}
+
+// RouteIterator allows iteration of the named routes in the store.
+type RouteIterator interface {
+	Valid() bool
+	Next() bool
+	Seek(string) bool
+	Error() error
+	Name() string
+	Route() *Route
+	Release()
+}
+
+type RouteNotFound struct {
+}
+
+var ErrRouteNotFound = errRouteNotFound()
```

Comment:
Probably don't need to have a custom error type here.
```
var ErrRouteNotFound = errors.New("route not found")
```

Reply:
Over in web/[web|api].go, there were type assertions made on the kind of LevelDB error that influenced the response Example at:

https://github.com/kellegous/go/blob/master/web/api.go#L136

I made a custom error type to allow any implementing backend the ability to return a common error type so the web portions of the code didn't have to have any knowledge of the backing store.

Reply:
I'll admit that I haven't used the 1.13 `errors` package near enough to feel super comfortable with it. But I thought that wrapping allowed the augmenting of existing value errors. So, for instance, we would have the following definition for `ErrRouteNotFound` in `internal`

```
var ErrorRouteNotFound = errors.new("route not found")
```

Then in all the places where we previously did this:

```
if err == leveldb.ErrNotFound { }
```

we would instead do this,

```
if errors.Is(err, internal.ErrRouteNotfound) { }
```

and finally, if a backend needs to return a custom error when a route is not found, it would use error wrapping.

So it would either create a custom error type and its `Wrap` method would return the underlying `error` which would be `internal.ErrRouteNotFound`. That should still cause `errors.Is(err, internal.ErrRouteNotFound)` to evaluate to `true`.

The other option would be for the backend to use `errors.Wrap` to wrap the existing `internal.ErrRouteNotFound` with additional string information. For instance,

```
return errors.Wrap("the file system is actually a dragon: %W", internal.ErrRouteNotFound)
```

Anyway, that's my understanding of 1.13 errors. Mostly from reading through this https://blog.golang.org/go1.13-errors a little while back. Is that consistent with what you have in mind?

Reply:
I think I read your original comment as "why return an error?" instead of "why have a custom type?"

You're right, I removed the custom type, it's not needed.